### Lambda Invocations

When invoking a function, it can either be a Synchronous or an Asynchronous invocation.

#### Synchronous Invocation:

- Lambda runs the function > waits for a result > and then returns response.

The service calling the function will know if the function completed successfully or not. Ie. API Gateway invoking a function and returning an error code to the caller.

#### Asynchronous Invocation:

- No acknowledgement to let you know that the invocation has successfully processes.

The service calling the function is not notified if the function failed to complete successfully. Ie. S3 invoking a function when an object is created.

So how does one handle errors when they occur in Asynchronous Invocations?

- Lambda Retries: If function returns error > Lambda automatically perform 2 retries.
- Common errors come from errors in the code, or a function timeout before the processing is complete.
- Lambda waits 1 minute before first retry, if that failes, it waits 2 minutes before the second retry.

#### Dead-Letter Queues(DLQs)

Save failed invocations for further processing. These are associated with a particular version of a function, and can even be the event source for another function or the same function, allowing one to re-process events.

These can only be used to handle failures only.

When creating DLQs we can use SQS to hold failed events in the queue until they are retreived. Or, we can use SNS to send notifications about failed events to one or multiple destinations.

#### Lambda Destinations

Can be used to configure Lambda to send invocation records to another service.

Lambda function runs:

- On success, a record is sent to SQS when the invocation is processed.
- On failure, notifications can be sent to a destination, notifying a service or person of all of the failed attempts.
